### PR TITLE
Sketch list pagination support for the API client

### PR DESCRIPTION
**Pull request by berggren**
This PR adds pagination support to the API client list_sketches()

```
list_sketches(scope='user', include_archived=True)
```

### CODE REVIEW

File: api_client/python/timesketch_api_client/client.py [hunk 1]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
-                sketch_id=sketch_id, api=self, sketch_name=sketch_name)
-            sketches.append(sketch_obj)
-        return sketches
+        url_params = {
+            'per_page': per_page,
+            'scope': scope,
+            'include_archived': include_archived
+        }
+        response = self.fetch_resource_data('sketches/', url_params)
+        meta = response['meta']
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:435]: can you use 

```
meta = response.get('meta', {})
```

File: api_client/python/timesketch_api_client/client.py [hunk 2]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+
+        total_pages = meta.get('total_pages', 0)
+        current_page = meta.get('current_page')
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:438]: maybe with a default value of 0

File: api_client/python/timesketch_api_client/client.py [hunk 3]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+        while current_page <= total_pages:
+            next_page = meta.get('next_page')
+            if not next_page:
+                next_page = meta.get('current_page')
+            url_params = {
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:443]: this is already set, no need to set it again, just change the value that is changed.


```
url_params['page'] = next_page
```

File: api_client/python/timesketch_api_client/client.py [hunk 4]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+                'per_page': per_page,
+                'page': next_page,
+                'scope': scope,
+                'include_archived': include_archived
+            }
+            response = self.fetch_resource_data('sketches/', params=url_params)
+            meta = response['meta']
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:450]: again.. get

File: api_client/python/timesketch_api_client/client.py [hunk 5]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
-            sketch_obj = sketch.Sketch(
-                sketch_id=sketch_id, api=self, sketch_name=sketch_name)
-            sketches.append(sketch_obj)
-        return sketches
+        url_params = {
+            'per_page': per_page,
+            'scope': scope,
+            'include_archived': include_archived
+        }
+        response = self.fetch_resource_data('sketches/', url_params)
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:434]: you fetch the response here, and you don't do anything witht he response (as in no yielding of sketches) until you go into the loop and then you start by overwriting this repsonse by a new one, is that the intended behavior? or want to go through this first repsonse first, before moving on to the next page?

File: api_client/python/timesketch_api_client/client.py [hunk 6]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+                next_page = meta.get('current_page')
+            url_params = {
+                'per_page': per_page,
+                'page': next_page,
+                'scope': scope,
+                'include_archived': include_archived
+            }
+            response = self.fetch_resource_data('sketches/', params=url_params)
+            meta = response['meta']
+            for sketch_dict in response['objects']:
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:451]: response.get('objects', [])

File: api_client/python/timesketch_api_client/client.py [hunk 7]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+        }
+        response = self.fetch_resource_data('sketches/', url_params)
+        meta = response['meta']
+
+        total_pages = meta.get('total_pages', 0)
+        current_page = meta.get('current_page')
+        while current_page <= total_pages:
+            next_page = meta.get('next_page')
+            if not next_page:
+                next_page = meta.get('current_page')
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:442]: so why is this?


also you use `current_page` somewhat, but here you are fetching the last current page from the meta... where you going to use the `current_page` or the meta['current_page']

File: api_client/python/timesketch_api_client/client.py [hunk 8]
```diff
@@ -414,21 +415,46 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+            }
+            response = self.fetch_resource_data('sketches/', params=url_params)
+            meta = response['meta']
+            for sketch_dict in response['objects']:
+                sketch_id = sketch_dict['id']
+                sketch_name = sketch_dict['name']
+                sketch_obj = sketch.Sketch(
+                    sketch_id=sketch_id, api=self, sketch_name=sketch_name)
+                yield sketch_obj
+            current_page += 1
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:457]: should the current_page not be set as `next_page` at this point?

**Review by kiddinn — CHANGES_REQUESTED** [review]
few comments, can you also increment the version number in the API client?

File: api_client/python/timesketch_api_client/client.py (hunk 1 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:435]: Done.

File: api_client/python/timesketch_api_client/client.py (hunk 2 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:438]: ack -setting to 1 as that is the start.

File: api_client/python/timesketch_api_client/client.py (hunk 3 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:443]: done.

File: api_client/python/timesketch_api_client/client.py (hunk 4 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:450]: done.

File: api_client/python/timesketch_api_client/client.py (hunk 6 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:448]: done

File: api_client/python/timesketch_api_client/client.py (hunk 7 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py]: Refactored this

File: api_client/python/timesketch_api_client/client.py (hunk 8 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py]: Refactored

File: api_client/python/timesketch_api_client/client.py [hunk 9]
```diff
@@ -414,21 +415,42 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
         return pandas.DataFrame(lines)
 
-    def list_sketches(self):
+    def list_sketches(self, per_page=50, scope='user', include_archived=True):
         """Get a list of all open sketches that the user has access to.
 
-        Returns:
-            List of Sketch objects instances.
+        Args:
+            per_page: Number of items per page when paginating.
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:422]: can you add the default values... eg

```
per_page (int): Optional number of items per page when paginating. Defaults to 50.
```

Something like that (and add the type there)

File: api_client/python/timesketch_api_client/client.py [hunk 10]
```diff
@@ -414,21 +415,42 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
+        has_next_page = True
+
+        while has_next_page:
+            url_params['page'] = page
+            response = self.fetch_resource_data('sketches/', params=url_params)
+            meta = response.get('meta', {})
+
+            page = meta.get('next_page')
+            if not page:
+                page = meta.get('current_page')
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:445]: you can just skip this assignment... not doing anything with `page` at this point...

File: timesketch/api/v1/resources/sketch.py [hunk 11]
```diff
@@ -66,11 +70,18 @@ def get(self):
...
         page = args.get('page', 1)
+        per_page = args.get('per_page', 20)
         search_query = args.get('search_query', None)
+        include_archived = args.get('include_archived', True)
 
-        if current_user.admin:
+        if not page:
+            page = 1
+
+        print(page)
```
Comment (kiddinn) [inline timesketch/api/v1/resources/sketch.py:82]: probably want to remove this ;)

**Review by kiddinn — COMMENTED** [review]
ok, just few more minor nits

File: timesketch/api/v1/resources/sketch.py (hunk 11 above)
Comment (berggren) [reply timesketch/api/v1/resources/sketch.py:82]: done

File: api_client/python/timesketch_api_client/client.py (hunk 9 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:422]: done

File: api_client/python/timesketch_api_client/client.py (hunk 10 above)
Comment (berggren) [reply api_client/python/timesketch_api_client/client.py:445]: good catch, removed

**Review by berggren — COMMENTED** [review]
PTAL

File: api_client/python/timesketch_api_client/client.py [hunk 12]
```diff
@@ -414,21 +415,41 @@ def get_aggregator_info(self, name='', as_pandas=False):
...
 
-    def list_sketches(self):
+    def list_sketches(self, per_page=50, scope='user', include_archived=True):
         """Get a list of all open sketches that the user has access to.
 
-        Returns:
-            List of Sketch objects instances.
+        Args:
+            per_page: Number of items per page when paginating. Default is 50.
+            scope: What scope to get sketches as (e.g user, admin, shared)
```
Comment (kiddinn) [inline api_client/python/timesketch_api_client/client.py:423]: can you add the default avalue and add that this is optional to all three items here?

**Review by kiddinn — APPROVED** [review]
minor nit on the docstring, otherwise LGTM

**Pull request merged**